botocore/client.py

```diff
@@ -913,6 +912,11 @@ def _make_api_call(self, operation_name, api_params):
             'has_streaming_input': operation_model.has_streaming_input,
             'auth_type': operation_model.auth_type,
         }
+        api_params = self._emit_api_params(
+            api_params=api_params,
+            operation_model=operation_model,
+            context=request_context,
+        )
```

Comment:
Maybe this could be called from within `_resolve_endpoint_ruleset` at the very start of it? I'm thinking in purely organizational terms just because `_make_api_call` is a rather long, complex function already.